src/scenarios/mauiandroid/pre.py

```diff
@@ -3,47 +3,58 @@
-parser.add_argument('--unzip', help='Unzip APK and report extracted tree', action='store_true', default=False)
+parser = ArgumentParser(add_help=False)
 parser.add_argument(
-        '--apk-name',
-        dest='apk',
-        required=True,
+        '-o', '--output',
+        dest='output_dir',
+        required=False,
         type=str,
-        help='Name of the APK to setup')
-args = parser.parse_args()
-
-if not os.path.exists(PUBDIR):
-    os.mkdir(PUBDIR)
-apkname = args.apk
-apknamezip = '%s.zip' % (apkname)
-if not os.path.exists(apkname):
-    getLogger().error('Cannot find %s' % (apkname))
-    exit(-1)
-if args.unzip:
-    if not os.path.exists(apknamezip):
-        copyfile(apkname, apknamezip)
-
-    with ZipFile(apknamezip) as zip:
-        zip.extractall(os.path.join('.', PUBDIR))
-
-    assets_dir = os.path.join(PUBDIR, 'assets')
-    assets_zip = os.path.join(assets_dir, 'assets.zip')
-    with ZipFile(assets_zip) as zip:
-        zip.extractall(assets_dir)
-
-    os.remove(assets_zip)
-else:
-    copyfile(apkname, os.path.join(PUBDIR, apkname))
-
-
-
+        help='capture of the output directory')
+args, unknown_args = parser.parse_known_args()
+
+# Download what we need
+with open ("MauiNuGet.config", "wb") as f:
+    f.write(requests.get(f'https://raw.githubusercontent.com/dotnet/maui/net7.0/NuGet.config', allow_redirects=True).content)
+
+precommands = PreCommands()
+precommands.install_workload('maui', ['--configfile', 'MauiNuGet.config'])
+
+# Setup the Maui folder
+precommands.new(template='maui',
+                output_dir=const.APPDIR,
+                bin_dir=const.BINDIR,
+                exename=EXENAME,
+                working_directory=sys.path[0],
+                no_restore=False)
+
+# Build the APK
+precommands.execute(['--no-restore', '--source', 'MauiNuGet.config'])
+
+# Remove the aab files as we don't need them, this saves space
+output_file_partial_path = f"com.companyname.{str.lower(EXENAME)}"
+if args.output_dir:
+    output_file_partial_path = os.path.join(args.output_dir, output_file_partial_path)
+
+os.remove(f"{output_file_partial_path}-Signed.aab")
+os.remove(f"{output_file_partial_path}.aab")
+
+# Copy the MauiVersion to a file so we have it on the machine
+result = subprocess.run(['powershell', '-Command', rf'Get-ChildItem .\{const.APPDIR}\obj\Release\net7.0-android\android-arm64\linked\Microsoft.Maui.dll | Select-Object -ExpandProperty VersionInfo | Select-Object ProductVersion | Select-Object -ExpandProperty ProductVersion'], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True)
+maui_version = result.stdout.decode('utf-8').strip()
+print(f'MAUI_VERSION: {maui_version}')
+if("sha" not in maui_version or "azdo" not in maui_version):
+    raise ValueError(f"MAUI_VERSION does not contain sha and azdo indicating failure to retrieve or set the value. MAUI_VERSION: {maui_version}")
+with open(f'{args.output_dir}/{MAUIVERSIONFILE}', 'w') as f:
+    f.write(maui_version)
```

Comment:
Is passing data via files the only approach for sending variables from pre to test? There isn't any other way we can pass arguments in or maybe use environment variables?

Reply:
Since the pre and test files are run on different machines, I did not find a good way to pass the version as a variable. Although I am open to suggestions and am looking to see if it would be possible to get an environment variable from the pre.py run, save it, and pass it later in a MSBuild task.

Reply:
Given the `proj` file is executed twice effectively, probably only sharing something through pipeline variable is an option (using `echo vso...`). But I don't think that's the way to go.

I think the file is OK. Maybe even create it as a JSON file, so it's ready for more values in the future (ultimate solution would be to have encapsulated helper for this).

Other option would be to encode the MAUI version into i.e. directory name (in the payload) and take it from there. Similar to what we do with report files in blob storage and all the different configurations. It's more hacky and limited than file, but also more straightforward.

I personally would vote for (JSON) file.

Reply:
I like this idea of having a generic Version JSON file. I will generalize this to use JSON, then down the line the same file can be used for any other version info we want to include from the build stage.

Reply:
For the current iteration, I added a shared set of version management functions to create a JSON file from a dictionary, and to read and save variables from this JSON file to the environment. This format minimizes the generation needed as the python json import provides most of the functionality. @caaavik-msft let me know if this seems like a reasonable middle ground.